### ISSUE TITLE
nixpkgs.cargo-watch fails to compile on macOS Catalina

### ISSUE THREAD

**Issue by rpearce** [NixOS/nixpkgs#120688]
**Describe the bug**
Trying to install `nixpkgs.cargo-watch` on macOS Catalina 10.15.7 fails with

```
   Compiling cargo-watch v7.6.1 (/private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source)
error: linking with `/nix/store/2pb1r2zgw6yiabv4v3d4j66d3vwgdd5q-clang-wrapper-7.1.0/bin/cc` failed: exit code: 1
  |
  = note: "/nix/store/2pb1r2zgw6yiabv4v3d4j66d3vwgdd5q-clang-wrapper-7.1.0/bin/cc" "-m64" "-arch" "x86_64" "-L" "/nix/store/akm781c35gr7mb14g5h0kqlk3cvyzmij-rustc-1.51.0/lib/rustlib/x86_64-apple-darwin/lib" "/private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/target/x86_64-apple-darwin/release/deps/cargo_watch.cargo_watch.6wybs3h6-cgu.0.rcgu.o" "-o" "/private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/target/x86_64-apple-darwin/release/deps/cargo_watch" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/target/x86_64-apple-darwin/release/deps" "-L" "/private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/target/release/deps" "-L" "/nix/store/akm781c35gr7mb14g5h0kqlk3cvyzmij-rustc-1.51.0/lib/rustlib/x86_64-apple-darwin/lib" "/nix/store/akm781c35gr7mb14g5h0kqlk3cvyzmij-rustc-1.51.0/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-94ac56476cd78265.rlib" "-framework" "CoreServices" "-framework" "CoreServices" "-liconv" "-lSystem" "-lresolv" "-lc" "-lm"
  = note: ld: library not found for -liconv
          clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
```

This looks very similar to https://github.com/NixOS/nixpkgs/pull/117559 and https://github.com/NixOS/nixpkgs/issues/36140...

**To Reproduce**

Option 1: `nix-env -i cargo-watch`

<details>
<summary>Option 2: try to build and then <code>nix develop</code> this flake on macOS Catalina</summary>

```nix
{
  description = "testing123";

  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/master";
    flake-utils = {
      url = "github:numtide/flake-utils";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    rust-overlay = {
      url = "github:oxalica/rust-overlay";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = { flake-utils, nixpkgs, rust-overlay, self }:
    flake-utils.lib.eachDefaultSystem (system:
      let
        overlays = [ (import rust-overlay) ];
        pkgs = import nixpkgs { inherit system overlays; };
        rust = pkgs.rust-bin.stable.latest.default;

      in rec {
        # start dummy stuff
        packages = flake-utils.lib.flattenTree { hello = pkgs.hello; };
        defaultPackage = packages.hello;
        apps.hello = flake-utils.lib.mkApp { drv = packages.hello; };
        defaultApp = apps.hello;
        # end dummy stuff

        devShell = pkgs.mkShell {
          buildInputs = [ pkgs.cargo-watch ];
          nativeBuildInputs = [ rust ];
        };
      }
    );
}
```

</details>

**Notify maintainers**
@xrelkd @ivan 

**Metadata**
<details>
<summary>Result of <code>nix-shell -p nix-info --run "nix-info -m"</code></summary>

```
 - system: `"x86_64-darwin"`
 - host os: `Darwin 19.6.0, macOS 10.15.7`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.4pre20210317_8a5203d`
 - channels(<user>): `"home-manager, nixpkgs-21.05pre278688.c0e88185200"`
 - nixpkgs: `/Users/<user>/.nix-defexpr/channels/nixpkgs`
```

</details>

**Comment by rpearce** [comment]
Taking after #117559 and #36140, I tried the following diff to my local `nixpkgs` and updated `nixpkgs.url` to `"path:/path/to/my/local/nixpkgs"`

```diff
diff --git a/pkgs/development/tools/rust/cargo-watch/default.nix b/pkgs/development/tools/rust/cargo-watch/default.nix
index 43f13b2e995..9831dc48b09 100644
--- a/pkgs/development/tools/rust/cargo-watch/default.nix
+++ b/pkgs/development/tools/rust/cargo-watch/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, lib, rustPlatform, fetchFromGitHub, CoreServices, rust }:
+{ stdenv, lib, rustPlatform, fetchFromGitHub, CoreServices, rust, libiconv }:
 
 rustPlatform.buildRustPackage rec {
   pname = "cargo-watch";
@@ -13,7 +13,7 @@ rustPlatform.buildRustPackage rec {
 
   cargoSha256 = "sha256-ku+tI0DIofV0EZ413sPjbJDUSqwTxiT8NWBeURrJW1k=";
 
-  buildInputs = lib.optional stdenv.isDarwin CoreServices;
+  buildInputs = lib.optionals stdenv.isDarwin [ CoreServices libiconv ];
 
   # `test with_cargo` tries to call cargo-watch as a cargo subcommand
   # (calling cargo-watch with command `cargo watch`)
```

but I then got this error

```
error: builder for '/nix/store/vhpbjh0lmf24a2jbrcbx68d2gvrsya38-cargo-watch-7.6.1.drv' failed with exit code 101;
       last 10 log lines:
       >
       > failures:
       >     with_announce
       >     with_error
       >     with_poll
       >     without_announce
       >
       > test result: FAILED(B. 0 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out; finished in 30.11s
       >
       > error: test failed, to rerun pass '--test echo'
       For full logs, run 'nix log /nix/store/vhpbjh0lmf24a2jbrcbx68d2gvrsya38-cargo-watch-7.6.1.drv'.
error: 1 dependencies of derivation '/nix/store/prsk6gf5g896qzah7zi6bh78bf0092fv-nix-shell-env.drv' failed to build
```

<details>
<summary>Here is the relevant full error log</summary>

```
running 4 tests
test with_poll ... FAILED
test without_announce ... FAILED
test with_announce ... FAILED
test with_error ... FAILED

failures:

---- with_poll stdout ----
thread 'with_poll' panicked at 'Unexpected failure.
code=<interrupted>
stderr=```"WARN - Polling for changes every 500 ms\n"```
code=<interrupted>
stdout=```""```
stderr=```"WARN - Polling for changes every 500 ms\n"```
', /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/cargo-watch-7.6.1-vendor.tar.gz/assert_cmd/src/assert.rs:149:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- without_announce stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: tests/snapshots/echo__without_announce.stdout.snap
Snapshot: without_announce.stdout
Source: tests/echo.rs:167
-old snapshot
+new results
────────────────────────────────────────────────────────────────────────────────
std_to_string(&mut main.stdout)
────────────┬───────────────────────────────────────────────────────────────────
    1       │-without announce
    2       │-
────────────┴───────────────────────────────────────────────────────────────────
stored new snapshot /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/tests/snapshots/echo__without_announce.stdout.snap.new
To update snapshots run `cargo insta review`
thread 'without_announce' panicked at 'snapshot assertion for 'without_announce.stdout' failed in line 167', /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/cargo-watch-7.6.1-vend


stored new snapshot /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/tests/snapshots/echo__without_announce.stdout.snap.new
To update snapshots run `cargo insta review`
thread 'without_announce' panicked at 'snapshot assertion for 'without_announce.stdout' failed in line 167', /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/cargo-watch-7.6.1-vend

---- with_announce stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: tests/snapshots/echo__with_announce.stdout.snap
Snapshot: with_announce.stdout
Source: tests/echo.rs:133
-old snapshot
+new results
────────────────────────────────────────────────────────────────────────────────
std_to_string(&mut main.stdout)
────────────┬───────────────────────────────────────────────────────────────────
    1       │-[Running 'echo with announce']
    2       │-with announce
    3       │-[Finished running. Exit status: 0]
    4       │-
────────────┴───────────────────────────────────────────────────────────────────
stored new snapshot /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/tests/snapshots/echo__with_announce.stdout.snap.new
To update snapshots run `cargo insta review`
thread 'with_announce' panicked at 'snapshot assertion for 'with_announce.stdout' failed in line 133', /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/cargo-watch-7.6.1-vendor.tar

---- with_error stdout ----
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot file: tests/snapshots/echo__with_error.stdout.snap
Snapshot: with_error.stdout
Source: tests/echo.rs:203
-old snapshot
+new results
────────────────────────────────────────────────────────────────────────────────
std_to_string(&mut main.stdout)
────────────┬───────────────────────────────────────────────────────────────────
    1       │-[Running 'echo with error && false']
    2       │-with error
    3       │-[Finished running. Exit status: 1]
    4       │-
────────────┴───────────────────────────────────────────────────────────────────
stored new snapshot /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/source/tests/snapshots/echo__with_error.stdout.snap.new
To update snapshots run `cargo insta review`
thread 'with_error' panicked at 'snapshot assertion for 'with_error.stdout' failed in line 203', /private/tmp/nix-build-cargo-watch-7.6.1.drv-1/cargo-watch-7.6.1-vendor.tar.gz/in


failures:
    with_announce
    with_error
    with_poll
    without_announce
```

</details>

**Comment by rpearce** [comment]
I also tried upgrading `cargo-watch` to the new [`v7.7.0`](https://github.com/passcod/cargo-watch/releases) with this derivation:

```nix
{ stdenv, lib, fetchFromGitHub
, rust, rustPlatform
# darwin dependencies
, CoreServices }:

rustPlatform.buildRustPackage rec {
  pname = "cargo-watch";
  version = "7.7.0";

  src = fetchFromGitHub {
    owner = "passcod";
    repo = pname;
    rev = "v${version}";
    sha256 = "sha256-z4FG69kJxCbkGn+k/bl9N2fYgp2NcYQjaNXC4m4pX4c=";
  };

  cargoSha256 = "sha256-mD8ZYEZJnsgl18MXE+P3NU/zhKm8zGkFj0PV0qjCfFw=";

  buildInputs = lib.optionals stdenv.isDarwin [ CoreServices];

  # `test with_cargo` tries to call cargo-watch as a cargo subcommand
  # (calling cargo-watch with command `cargo watch`)
  preCheck = ''
    export PATH="$(pwd)/target/${rust.toRustTarget stdenv.hostPlatform}/release:$PATH"
  '';

  meta = with lib; {
    description = "A Cargo subcommand for watching over Cargo project's source";
    homepage = "https://github.com/passcod/cargo-watch";
    license = licenses.cc0;
    maintainers = with maintainers; [ xrelkd ivan ];
  };
}
```

with no luck (the same `liconv` error), and I also tried `libiconv` with `v7.7.0`, as well:

```nix
{ stdenv, lib, fetchFromGitHub
, rust, rustPlatform
# darwin dependencies
, CoreServices, libiconv }:

rustPlatform.buildRustPackage rec {
  pname = "cargo-watch";
  version = "7.7.0";

  src = fetchFromGitHub {
    owner = "passcod";
    repo = pname;
    rev = "v${version}";
    sha256 = "sha256-z4FG69kJxCbkGn+k/bl9N2fYgp2NcYQjaNXC4m4pX4c=";
  };

  cargoSha256 = "sha256-mD8ZYEZJnsgl18MXE+P3NU/zhKm8zGkFj0PV0qjCfFw=";

  buildInputs = lib.optionals stdenv.isDarwin [ CoreServices libiconv ];

  # `test with_cargo` tries to call cargo-watch as a cargo subcommand
  # (calling cargo-watch with command `cargo watch`)
  preCheck = ''
    export PATH="$(pwd)/target/${rust.toRustTarget stdenv.hostPlatform}/release:$PATH"
  '';

  meta = with lib; {
    description = "A Cargo subcommand for watching over Cargo project's source";
    homepage = "https://github.com/passcod/cargo-watch";
    license = licenses.cc0;
    maintainers = with maintainers; [ xrelkd ivan ];
  };
}
```

but I got the same snapshot test failures.

**Comment by rpearce** [comment]
Looks like this might be relevant to this issue: https://github.com/passcod/cargo-watch/issues/170

**Comment by davidarmstronglewis** [comment]
Have a PR that I think addresses this here: https://github.com/NixOS/nixpkgs/pull/120953

**Comment by prusnak** [comment]
Fixed via #121347 and #121974
